src/node_id.rs

```diff
@@ -0,0 +1,334 @@
+use crate::node::NodeInfo;
+use bs58;
+use uint::*;
+use disco::{hash, DiscoHash};
+use std::{
+    cmp::Ordering,
+    convert::{TryFrom, TryInto},
+    fmt,
+    str::FromStr,
+};
+use time::{Duration, SteadyTime};
+
+construct_uint! {
+    /// 256-bit unsigned integer.
+    pub struct U256(4);
+}
+
+/// NodeId
+///
+/// fn generate() { disco::hash(public_key) }
+#[derive(Clone, Eq)]
+pub struct NodeId {
+    pub discohash: Vec<u8>,
+}
+
+impl fmt::Debug for NodeId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("NodeId").field(&self.to_base58()).finish()
+    }
+}
+
+impl fmt::Display for NodeId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.to_base58().fmt(f)
+    }
+}
+
+impl NodeId {
+    /// Builds a `NodeId` from a public key.
+    #[inline]
+    pub fn from_public_key(key: PublicKey) -> Result<NodeId, NodeIdGenError> {
+        let keyhash = hash(key.as_bytes(), 32);
+        let new_node = NodeId { discohash: keyhash };
+        if new_node.is_zero() {
+            return Err(NodeIdGenError::PubkeyHashZero);
+        } else {
+            return Ok(new_node);
+        }
+    }
+
+    /// Generate NodeId
+    ///
+    /// Default no hard mechanism for slowing id generation
+    #[inline]
+    pub fn generate() -> Result<NodeId, NodeIdGenError> {
+        // TODO: private key must be stored somewhere for signing messages?
+        let key = Keypair::generate(&mut rand::thread_rng());
+        let keyhash = hash(key.public.as_bytes(), 32);
+        let new_node = NodeId { discohash: keyhash };
+        if new_node.is_zero() {
+            return Err(NodeIdGenError::PubkeyHashZero);
+        } else {
+            return Ok(new_node);
+        }
+    }
+
+    /// Generate NodeId with Resistance
+    ///
+    /// Requires disco::hash(public_key) to be have `difficulty` number of trailing zeros
+    /// TODO: hash should be of some changing shared state for better sybil resistance
+    pub fn hard_generate(difficulty: usize, timeout: usize) -> Result<NodeId, NodeIdGenError> {
+        let clock = SteadyTime::now();
+        loop {
+            let new_id = NodeId::generate()?;
+            // default trailing zeros (remove `rev()` for leading zeros)
+            //let disco_iter = new_id.discohash.iter().rev();
+            let mut success = true;
+            for i in 0..difficulty {
+                if new_id.discohash.get(i).unwrap() != &0u8 {
+                    success = false;
+                }
+            }
+            if success {
+                return Ok(new_id);
+            }
+            //                           converts usize into i64 and panics if doesn't fit
+            // TODO: consider if this is best or if I should just pass i64 as an argument
+            if SteadyTime::now() - clock > Duration::seconds(timeout.try_into().unwrap()) {
+                return Err(NodeIdGenError::HardGenTimeOut);
```

Comment:
That's fine, I really dislike this method anyway. It's a placeholder for what I am thinking will be a more robust anti-sybil mechanism that uses blockchain consensus to manage the set of identities.

Reply:
You need a dht to build a blockchain. It's a chicken and egg problem. Think of this as being a building block for building a blockchain, not a blockchain.

Reply:
You don't need to use this DHT to build a blockchain. You can use a blockchain as a primitive. It's not a chicken or egg problem. Just because every blockchain has a DHT doesn't mean that a DHT can't use a blockchain for managing the set of nodes. It doesn't have to be (and probably shouldn't be) a blockchain that uses this DHT.

Reply:
so you're suggesting to use substrate to build this DHT?

Reply:
I'm suggesting using substrate to limit `NodeId` generation for this DHT.

Reply:
@troublescooter has some thoughts on how to build Sybil resistant dhts

Reply:
Can we open an issue for this? Depending on substrate is not something that we should decide without evaluating other options

Reply:
#9